diff --git a/mm/kasan/report.c b/mm/kasan/report.c
index 03a443579386..fa777489668b 100644
--- a/mm/kasan/report.c
+++ b/mm/kasan/report.c
@@ -292,6 +292,12 @@ void __kasan_report(unsigned long addr, size_t size, bool is_write, unsigned lon
 	if (likely(!report_enabled()))
 		return;
 
+	if (!is_write) {
+		clear_bit(KASAN_BIT_REPORTED, &kasan_flags);
+		printk(KERN_WARNING "?!?MAGIC?!?read->%llx size->%d", addr, size);
+		return;
+	}
+
 	disable_trace_on_warning();
 
 	tagged_addr = (void *)addr;
